### PR TITLE
Homework2: Aliaksei Uhalnikau

### PR DESCRIPTION
**Pull request by alexugalek**
New branch with commits only of homework 2 tasks

### CODE REVIEW

File: src/Homework2New/task_24.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+def eight(func=None):
+    return func(8) if func else 8
+
+
+def nine(func=None):
+    return func(9) if func else 9
+
+
+def plus(num):
+    return lambda x: x + num
```
Comment (belovmd) [inline src/Homework2New/task_24.py:54]: lamdba is best solution here

File: src/Homework2New/task_26.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    def __init__(self, value, message):
+        self._value = value
+        self._message = message
+
+    def __str__(self):
+        return f'Operation error {self._value}: {self._message}'
+
+    @property
+    def _type(self):
```
Comment (belovmd) [inline src/Homework2New/task_26.py:11]: it's unclear why do we need this property

File: src/Homework2New/task_26.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                    temp_result = OPERATORS[char](x1, x2)
+                    temp_stack_value.append(temp_result)
+
+                except ZeroDivisionError:
+                    return MyException(test_s, 'Division by zero')
+
+        if len(temp_stack_str) != 1:
+            return MyException(test_s, 'Wrong format')
+
+        expr = temp_stack_str[0][1:-1]
```
Comment (belovmd) [inline src/Homework2New/task_26.py:55]: indexing and slicing stack is not the best idea
stack should be used with pop and push (append) functions and that's all
also it's unclear why do we need str in list name :)

File: src/Homework2New/task_28.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    flag_1, flag_2, flag_3 = False, False, False
+    if len(data) >= 10:
+        for char in data:
+            if '0' <= char <= '9':
+                flag_1 = True
+            elif 'a' <= char <= 'z':
+                flag_2 = True
+            elif 'A' <= char <= 'Z':
+                flag_3 = True
+        return flag_1 == flag_2 == flag_3
```
Comment (belovmd) [inline src/Homework2New/task_28.py:19]: and and and?

File: src/Homework2New/task_29.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    :param text: string type
+    :return: string type
+    """
+
+    alphabet = 'qwertyuiopasdfghjklzxcvbnm'
+    count_dict = dict.fromkeys(alphabet, 0)
+    for char in text.lower():
+        if char in alphabet:
+            count_dict[char] += 1
+    return sorted(count_dict, key=lambda x: (-count_dict[x], x))[0]
```
Comment (belovmd) [inline src/Homework2New/task_29.py:23]: pretty implicit line

File: src/Homework2New/task_30.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    """Time formatter
+
+    :param time: string format of time '00:00'
+    :return: string format of time '12:00 a.m.'
+    """
+    if time[:2] == '00':
+        return '12' + time[2:] + ' a.m.'
+    elif time[:2] < '12':
+        return time[1:] + ' a.m.' if time[:2] < '10' else time + ' a.m.'
+    return time + ' p.m.' if time[:2] == '12' \
```
Comment (belovmd) [inline src/Homework2New/task_30.py:11]: ternary operators shouldn't be used if they are multi-line

File: src/Homework2New/task_32.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    Sort the given iterable so that its elements end up in the decreasing
+    frequency order, that is, the number of times they appear in elements
+    :param items: list of elements of string or integer type
+    :return: sorted list with elements frequency key
+    """
+    dict_count, result = dict.fromkeys(items, 0), []
+    for item in items:
+        dict_count[item] += 1
+    sorted_items = sorted(dict_count,
```
Comment (belovmd) [inline src/Homework2New/task_32.py:12]: `lst = dict_count.values(); lst.sort; return lst[::-1]?`

File: src/Homework2New/task_36.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                        'g': 7,
+                        'h': 8}
+
+    converted_board = [[letter_converter[element[0]],
+                        int(element[-1])] for element in pawns]
+
+    counter = 0
+    for ch_pawn in converted_board:
+        for pawn in converted_board:
+            if ch_pawn[1] == pawn[1] + 1 and \
```
Comment (belovmd) [inline src/Homework2New/task_36.py:23]: pretty unclear implementation )) too many numbers and operations

File: src/Homework2New/task_4.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    Return the words of the initial song that
+    Polycarpus used to make a dubsteb remix (WUB). Separate
+    the words with a space.
+    Examples
+    song_decoder("WUBWEWUBAREWUBWUBTHEWUBCHAMPIONSWUBMYWUBFRIENDWUB")
+    # =>  WE ARE THE CHAMPIONS MY FRIEND
+    """
+    list_of_words = song.split('WUB')
+    return ' '.join(filter(lambda x: x != '', list_of_words))
+    # return ' '.join([word for word in list_of_words if word])  Equal result
```
Comment (belovmd) [inline src/Homework2New/task_4.py:18]: debug comments

File: src/Homework2New/task_45.py [hunk 10]
```diff
@@ -0,0 +1,834 @@
...
+    def fight(self, army_1, army_2):
+        """Class method
+
+        Init fight between two armies while summary health one of them > 0
+        :param army_1: list of objects of Warrior or Knight class
+        :param army_2: list of object of Warrior or Knight class
+        :return: army_1.units[-1].is_alive property
+        """
+
+        unit_army1_nmb = unit_army2_nmb = 0
```
Comment (belovmd) [inline src/Homework2New/task_45.py:195]: this function looks too complex. it seems all this class can be implemented easier. 
you can not change it and use in further homework. (we'll have some for emulation with classes)

File: src/Homework2New/task_45.py [hunk 11]
```diff
@@ -0,0 +1,834 @@
...
+        """
+        if unit_b.__class__.__name__ == 'Defender':
+            attack = unit_a.attack - unit_b.defense \
+                if unit_a.attack > unit_b.defense else 0
+        else:
+            attack = unit_a.attack
+        unit_b.health -= attack
+        return attack
+
+    def vampire_recovery(unit, attack):
```
Comment (belovmd) [inline src/Homework2New/task_45.py:282]: all these methods should be put to according models

File: src/Homework2New/task_45.py [hunk 12]
```diff
@@ -0,0 +1,834 @@
...
+            while nmb < len(army_2.units):
+                if not army_2.units[nmb].is_alive:
+                    army_2.units.pop(nmb)
+                else:
+                    nmb += 1
+
+        return army_1.units != []
+
+
+def fight(unit_1, unit_2, unit_1_2=None, unit_2_2=None):
```
Comment (belovmd) [inline src/Homework2New/task_45.py:257]: we should move it to battle class,

File: src/Homework2New/task_52.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
+def checkio(expression):
```
Comment (belovmd) [inline src/Homework2New/task_52.py:1]: this task is often used during interview. please don't forget your stack solution.

File: src/Homework2New/task_55.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+    def set_font(self, font=''):
+        """Set font in text"""
+        if font:
+            self.font = font
+
+    def show(self):
+        """Show new text"""
+        if self.font:
+            return f'[{self.font}]{self.text}[{self.font}]'
```
Comment (belovmd) [inline src/Homework2New/task_55.py:33]: these f-strings ^)

File: src/Homework2New/task_8.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    a string to a new string where each character
+    in the new string is "(" if that character
+    appears only once in the original string, or ")"
+    if that character appears more than once in the
+    original string. Ignore capitalization when
+    determining if a character is a duplicate.
+    """
+    output_string = ''
+    test_dict = {}
+    # Create a dict with keys consists from word chars
```
Comment (belovmd) [inline src/Homework2New/task_8.py:14]: redundant comments

File: src/Homework2New/task_11.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
+def find_outlier(integers):
+    """Instructions
```
Comment (belovmd) [inline src/Homework2New/task_11.py:2]: missed full task description

File: src/Homework2New/task_15.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    """Instructions
+
+    Given two strings s1 and s2, we want to visualize
+    how different the two strings are. We will only
+    take into account the lowercase letters (a to z).
+    First let us count the frequency of each lowercase
+    letters in s1 and s2.
+    """
+    s = 'qwertyuioplkjhgfdsazxcvbnm'
+    list_result = []
```
Comment (belovmd) [inline src/Homework2New/task_15.py:11]: it seems this solution can be re-writed a bit more clear

**Review by belovmd — CHANGES_REQUESTED** [review]
Thank you for your work. It's great. 
Please address next paragraphs from review notes [Review Notes](https://github.com/belovmd/it-academy-python-spring/blob/master/homework_review/Homework2.md):
- 8
- 14
- 18

File: src/Homework2New/task_28.py (hunk 4 above)
Comment (alexugalek) [reply src/Homework2New/task_28.py:19]: My bad.

File: src/Homework2New/task_30.py (hunk 6 above)
Comment (alexugalek) [reply src/Homework2New/task_30.py:11]: Got it

**Review by belovmd — COMMENTED** [review]
10

**Pull request closed without merging**